respx/models.py

```diff
@@ -65,14 +64,42 @@
 
 DefaultType = TypeVar("DefaultType", bound=Any)
 
-Regex = type(re.compile(""))
 Kwargs = Dict[str, Any]
-URLPatternTypes = Union[str, Pattern[str], URL]
+URLPatternTypes = Union[str, Pattern[str], URL, httpx.URL]
 JSONTypes = Union[str, List, Dict]
 ContentDataTypes = Union[bytes, str, JSONTypes, Callable, Exception]
+QueryParamTypes = Union[bytes, str, List[Tuple[str, Any]], Dict[str, Any]]
 
-istype = lambda t, o: isinstance(o, t)
-isregex = partial(istype, Regex)
+
+def build_url(
+    url: URLPatternTypes, *, base: str = "", params: Optional[QueryParamTypes] = None
+) -> Union[httpx.URL, Pattern[str]]:
+    if not url:
+        if params is not None:
+            raise ValueError("Params cannot be used with empty url.")
+        return None
+
+    if isinstance(url, Pattern):
+        if params is not None:
+            if r"\?" in url.pattern and params is not None:
+                raise ValueError(
+                    "Request url pattern contains a query string, which is not "
+                    "supported in conjuction with params argument."
+                )
+            query_params = str(httpx.QueryParams(params))
+            url = re.compile(url.pattern + re.escape(fr"?{query_params}"))
+        return re.compile(urljoin(base, url.pattern))
+    if isinstance(url, str):
+        url = urljoin(base, url)
```

Comment:
This part also need support if the given url is a `httpx.URL`.

I think we need to..
1. create `httpx.URL` from str, *if string*
2. create base `httpx.URL` and ensure [trailing slash](https://github.com/encode/httpx/blob/master/httpx/_client.py#L96)
3. merge [base with url](https://github.com/encode/httpx/blob/master/httpx/_client.py#L306)
4. merge [params](https://github.com/encode/httpx/blob/master/httpx/_client.py#L309)